### PR TITLE
MRG: More work on Corrmap

### PR DESCRIPTION
**Pull request by jona-sassenhagen**
See #1985 and discussion there.

I've only fixed a few minor things that went broke with the streamlining done by @dengemann and @Eric89GXL.

Passes tests on my machine, and works on my sample data.

Anybody wanna give it a try?

### CODE REVIEW

**Comment by jona-sassenhagen** [comment]
@Eric89GXL We currently have keywords ```plot``` and ```show```. Shouldn't it be only ```show```, or am I missing the rationale behind what you did there?

**Comment by Eric89GXL** [comment]
IIRC someone at some point wanted the functionality without any plots being made, thus the `plot` option. Even doing the plots without showing them could incur some overhead. This suggests to me that the two bits of functionality should be separated (promotes orthogonality of the functions), but I'm not actually very familiar with what it does.

**Comment by jona-sassenhagen** [comment]
I guess there could be a situation where you want the figures constructed, but only plotted later, in a different context, but right now, to me it seems mostly confusing to have both keywords.

**Comment by agramfort** [comment]
the rationale is:

- if function / method starts with plot_ then we need a show param.
- if it's a function that runs numerics and can plot something for visual
inspection then use a plot param

ok?

**Comment by jona-sassenhagen** [comment]
I'm repeatedly coming across situations where I'd like to tell other users of MNE to use corrmap. I forgot, did we have any reservations regarding merging this once cleaned up, and which? It is after all an experimentally verified algorithm. @agramfort

**Comment by dengemann** [comment]
No I don't think so. We can merge it. Can you rebase?

**Comment by Eric89GXL** [comment]
@jona-sassenhagen did the points about plotting and showing get cleared up through your changes?

**Comment by agramfort** [comment]
please indeed rebase / clean up. thx

**Comment by jona-sassenhagen** [comment]
@Eric89GXL from what I understand, there are a bunch of changes here made by you that are seemingly unrelated to corrmap, e.g. the ones in mne/cov.py or mne/tests/test_import_nesting.py. I probably don't understand all of them. I'd like to continue working only with code I actually roughly understand the implications of, would it be possible to stash all of the changes irrelevant to corrmap?

**Comment by Eric89GXL** [comment]
The `test_import_nesting.py` changes were necessary in order to get tests to pass on this branch, so they need to stay. The changes to `cov.py` are pretty minor (overriding `copy` method) so I wouldn't worry about that one. I think if you don't touch either of those changesets you should be fine to continue working on the other bits.

**Comment by jona-sassenhagen** [comment]
There we are. I didn't really do anything but catch up with master.

**Comment by jona-sassenhagen** [comment]
Merge please? :)

**Comment by Eric89GXL** [comment]
@jona-sassenhagen for future reference we prefer rebasing to merging. I'll let @dengemann or @agramfort have a look and merge if they're happy.

**Comment by jona-sassenhagen** [comment]
There is 1 error in one of the Travis checks:

	ERROR: Test fieldtrip_client
	----------------------------------------------------------------------
	Traceback (most recent call last):
	  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
		self.test(*self.arg)
	  File "/home/travis/build/mne-tools/mne-python/mne/utils.py", line 622, in dec
		return function(*args, **kwargs)
	  File "/home/travis/build/mne-tools/mne-python/mne/realtime/tests/test_fieldtrip_client.py", line 55, in test_fieldtrip_client
		tmax=5, wait_max=1) as rt_client:
	  File "/home/travis/build/mne-tools/mne-python/mne/realtime/fieldtrip_client.py", line 91, in __enter__
		raise RuntimeError('Could not connect to FieldTrip Buffer')
	RuntimeError: Could not connect to FieldTrip Buffer

Rest is okay. I think the error is not about the actual code?

File: doc/whats_new.rst [hunk 1]
```diff
@@ -196,10 +196,15 @@ Changelog
 
    - Add ``set_montage`` method to the ``Raw``, ``Epochs``, and ``Evoked`` objects by `Teon Brooks`_ and `Denis Engemann`_
 
+   - Add ``DigMontage`` class and reader to interface with digitization info by `Teon Brooks`_ and `$
+
+   - Add ``set_montage`` method to the ``Raw``, ``Epochs``, and ``Evoked`` objects by `Teon Brooks` $
```
Comment (agramfort) [inline doc/whats_new.rst]: ???

File: examples/preprocessing/plot_corrmap_detection.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
+"""
+==========================================================
+Identify similar ICs across multiple datasets via CORRMAP
+==========================================================
```
Comment (agramfort) [inline examples/preprocessing/plot_corrmap_detection.py]: insert empty line here

File: mne/cov.py [hunk 3]
```diff
@@ -154,6 +157,16 @@ def save(self, fname):
...
 
+    def copy(self):
+        """Copy the Covariance object
+
+        Returns
+        -------
+        cov : instance of Covariance
+            The copied object.
+        """
+        return deepcopy(self)
```
Comment (agramfort) [inline mne/cov.py]: this should have been done in another PR.

File: examples/preprocessing/plot_corrmap_detection.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+icas = [ICA(n_components=20, random_state=0).fit(epochs)
+        for epochs in all_epochs]
+
+# 2) Use corrmap to identify the maps best corresponding
+#    to a pre-specified template across all subsets
+#    (or, in the real world, multiple participant data sets)
+
+template = (0, 0)
+fig_template, fig_detected = corrmap(icas, template=template, label="blinks")
```
Comment (agramfort) [inline examples/preprocessing/plot_corrmap_detection.py]: corrmap plots something?

Functions that do plots usually start with plot_ or otherwise take a show=True|False param.

**Comment by agramfort** [comment]
@jona-sassenhagen you should squash all commits into one to make sure the diff is up to date and history is clean.

File: examples/preprocessing/plot_corrmap_detection.py (hunk 4 above)
Comment (jona-sassenhagen) [inline examples/preprocessing/plot_corrmap_detection.py]: It has the side effect of showing a diagnostics plot, but its primary purpose is modifying ICA objects.

File: mne/cov.py (hunk 3 above)
Comment (Eric89GXL) [inline mne/cov.py]: Sorry that's my bad. Feel free to remove. I think I added it because there
were a lot of copies being done.

**Comment by Eric89GXL** [comment]
Squash by author though so blame works, right @agramfort?

**Comment by agramfort** [comment]
If you can. Or everything in one if its a pain

**Comment by agramfort** [comment]
> It has the side effect of showing a diagnostics plot, but its primary purpose is modifying ICA objects.

ok. In the example please make the show=True param explicit.

**Comment by jona-sassenhagen** [comment]
Soooo ... what exactly do I do to squash? 

    git reset --soft HEAD~3000
    git commit -m "squash"

?

This is what Google tells me.

**Comment by agramfort** [comment]
have a look at
http://blog.houen.net/git-large-rebase-made-simple-easy-painless/

**Comment by jona-sassenhagen** [comment]
Ok so I create a new branch, merge in this branch, reset, merge, and then make a new PR for it?

**Comment by agramfort** [comment]
Don't open a new PR. Just force push

**Comment by jona-sassenhagen** [comment]
I think that was it? Thanks @agramfort

File: mne/utils.py [hunk 5]
```diff
@@ -1872,3 +1871,19 @@ def random_permutation(n_samples, random_state=None):
...
+    """Compute pearson correlations between a vector and a matrix"""
+    if len(x) == 0 or len(y) == 0:
+        raise ValueError('x or y has zero length')
+    fast_dot = _get_fast_dot()
+    X = np.array(x, float)
+    Y = np.array(y, float)
+    X -= X.mean(0)
+    Y -= Y.mean(0)
+    x_sd = X.std(0, ddof=1)
+    # if covariance matrix is fully expanded, Y needs a transpose / brodcasting
```
Comment (agramfort) [inline mne/utils.py]: brodcasting -> broadcasting

File: mne/tests/test_utils.py [hunk 6]
```diff
@@ -414,6 +416,27 @@ def test_check_type_picks():
...
+                  [7.46, 6.77, 12.74, 7.11, 7.81, 8.84,
+                   6.08, 5.39, 8.15, 6.42, 5.73],
+                  [8, 8, 8, 8, 8, 8, 8, 19, 8, 8, 8],
+                  [6.58, 5.76, 7.71, 8.84, 8.47, 7.04,
+                   5.25, 12.50, 5.56, 7.91, 6.89]])
+
+    r = compute_corr(x, y.T)
+    r2 = np.array([np.corrcoef(x, y[i])[0, 1]
+                   for i in range(len(y))])
+    assert_allclose(r, r2, atol=1e-3)
```
Comment (agramfort) [inline mne/tests/test_utils.py]: 1e-3 ? this seems high.

**Comment by jona-sassenhagen** [comment]
Done.

**Comment by Eric89GXL** [comment]
@jona-sassenhagen it looks like your squash commit, and your latest commit (but not the three in between) don't link to your user account properly. Perhaps you are developing on two machines/logins where one does not have the git user info configured improperly?

**Comment by jona-sassenhagen** [comment]
Yes, I've been using an alternative computer for some of these, is that a problem?

**Comment by Eric89GXL** [comment]
It will make it so things like `git blame` and GitHub links don't work properly, so it would be good but not critical if you could fix it.

**Comment by Eric89GXL** [comment]
You can develop in as many places as you want so long as the git user information is configured properly in all those places.

**Comment by jona-sassenhagen** [comment]
Ok I'm setting it up on all computers I could potentially use, is there a simple way to make the already pushed stuff point to me?

**Comment by Eric89GXL** [comment]
You can re-squash if you want. Now that the merges are gone you can do it with `git rebase -i HEAD~5`, and then set all but the first commit to be `s` or `squash` instead of `pick`, and the latter 4 commits will be combined with the first. Then you can push force. As usual I recommend backing up your branch before doing that, though :)

**Comment by jona-sassenhagen** [comment]
Nah let's do it as-is :)

**Comment by jona-sassenhagen** [comment]
Sorry I'm still terrible at git.

**Comment by Eric89GXL** [comment]
@jona-sassenhagen it looks like there is something wrong, there is a conflict and there is a merge that shouldn't be there :(

**Comment by jona-sassenhagen** [comment]
... Can you maybe take over from the last commit before what I did today? There was a merge conflict, but it was just whats_new. Sorry.

**Comment by Eric89GXL** [comment]
Which commits are the right ones to take? I can't quite tell from the history.

**Comment by Eric89GXL** [comment]
Or should I just re-squash all the changes?

**Comment by jona-sassenhagen** [comment]
70d3895 is the one - just whats_new needs a fix.

**Comment by Eric89GXL** [comment]
Hmm, cherry-picking that one I get a conflicts outside of `whats_new.py`, including in `ica`. Is that expected?

**Comment by jona-sassenhagen** [comment]
When I checkout 70d3895 and merge in master now, I get a merge conflict in whats_new and in mne/tests/test_utils.py. The test_utils one should be simple: the import from mne.utils in line 10-20 shoudl be

```
from mne.utils import (set_log_level, set_log_file, _TempDir,
                       get_config, set_config, deprecated, _fetch_file,
                       sum_squared, estimate_rank,
                       _url_to_local_path, sizeof_fmt, _check_subject,
                       _check_type_picks, object_hash, object_diff,
                       requires_good_network, run_tests_if_main, md5sum,
                       ArgvSetter, _memory_usage, check_random_state,
                       _check_mayavi_version, requires_mayavi,
                       set_memmap_min_size, _get_stim_channel, _check_fname,
                       create_slices, _time_mask, random_permutation,
                      get_call_line, verbose, compute_corr)
```

ica.py is merged in without problems.

**Comment by Eric89GXL** [comment]
Ohhh, it's not just that one commit, it's that commit and the ones that precede it. I thought you had squashed it all down to one.

**Comment by jona-sassenhagen** [comment]
Sorry, no. That's what I mean - everything up to 70d3895.

**Comment by Eric89GXL** [comment]
@jona-sassenhagen I have the new version up on my fork `Eric89GXL` under the branch `corrmap`. If you do something like this, it will update this PR:
```
git checkout patch-8
git fetch Eric89GXL
git reset --hard Eric89GXL/corrmap
git push origin --force patch-8
```
Or you can give me push access to your repo and I'll push it there.

**Comment by Eric89GXL** [comment]
@jona-sassenhagen pushed, we'll see if the CIs are happy. Ready to go from your end if it is?

**Comment by jona-sassenhagen** [comment]
Yup. I can't think of a use case I haven't covered. Thanks.

**Comment by Eric89GXL** [comment]
@agramfort or @dengemann did either of you want to look again, or okay for merge from you?

**Comment by Eric89GXL** [comment]
@jona-sassenhagen Travis fails because of a couple of style things, I'll push a fix

**Comment by Eric89GXL** [comment]
CIs are happy, +1 for merge from my end.

**Comment by jona-sassenhagen** [comment]
Nice.

**Comment by agramfort** [comment]
thanks guys !

**Comment by jona-sassenhagen** [comment]
Thanks!

**Comment by teonlamont** [comment]
thanks @jona-sassenhagen! looking forward to using it

**Comment by teonlamont** [comment]
and @Eric89GXL :)

**Comment by dengemann** [comment]
Amazing, finally merged. Thanks @jona-sassenhagen !

File: mne/preprocessing/ica.py [hunk 7]
```diff
@@ -2141,3 +2168,286 @@ def _band_pass_filter(ica, sources, target, l_freq, h_freq, verbose=None):
...
+    fast_dot = _get_fast_dot()
+    if components is None:
+        components = list(range(ica.n_components_))
+    maps = fast_dot(ica.mixing_matrix_[:, components].T,
+                    ica.pca_components_[:ica.n_components_])
+    return maps
+
+
+def _find_max_corrs(all_maps, target, threshold):
+    """Compute correlations betwen template and target components"""
```
Comment (teonlamont) [inline mne/preprocessing/ica.py]: betwen --> between

Comment (agramfort) [inline mne/preprocessing/ica.py]: please fix directly in master these types of things.

thx
